diag_manager/fms_diag_yaml_object.F90

```diff
@@ -33,32 +33,24 @@ module fms_diag_yaml_object_mod
-integer, parameter :: NUM_REGION_ARRAY = 8
-  !> @brief The files type matching a C struct containing diag_yaml information
-    !> @ingroup fms_diag_files_mod
-type, bind(c) :: diag_yaml_files_struct
-     character (kind=c_char) :: file_fname (20) !< file name
-     character (kind=c_char) :: file_frequnit (7) !< the frequency unit
-     integer (c_int)    :: file_freq !< the frequency of data
-     character (kind=c_char) :: file_timeunit(7) !< The unit of time
-     character (kind=c_char) :: file_unlimdim(8) !< The name of the unlimited dimension
-     character (kind=c_char) :: file_write (5) !< false if the user doesn’t want the file to be 
-                                               !! created (default is true).
-     character (kind=c_char) :: file_realm (3) !< The modeling realm that the variables come from
-     real (c_float) :: file_region (NUM_REGION_ARRAY) !< Bounds of the regional section to capture
-     integer (c_int) :: file_new_file_freq !< Frequency for closing the existing file
-     character (kind=c_char) :: file_new_file_freq_units (3)!< Time units for creating a new file. 
-                                                        !! Required if “new_file_freq” used
-     integer (c_int) :: file_start_time !< Time to start the file for the first time. Requires “new_file_freq”
-     integer (c_int) :: file_duration !< How long the file should receive data after start time 
-                                      !! in “file_duration_units”.  This optional field can only 
-                                      !! be used if the start_time field is present.  If this field
-                                      !! is absent, then the file duration will be equal to the
-                                      !! frequency for creating new files.  
-                                      !! NOTE: The file_duration_units field must also be present if
-                                      !! this field is present.
-    character (kind=c_char) :: file_duration_units (3)!< The file duration units
-end type diag_yaml_files_struct
+integer, parameter :: NUM_SUB_REGION_ARRAY = 8
+integer, parameter :: MAX_STR_LEN = 255
+
+!> @brief type to hold the sub region information about a file
+type sub_region_type
+     character (len=:), allocatable :: grid_type !< Flag indicating the type of region,
+                                                 !! acceptable values are "latlon" and "index"
+     real :: lat_lon_sub_region (NUM_SUB_REGION_ARRAY)  !< Array that stores the grid point bounds for the sub region
+                                                        !! to use if grid_type is set to "latlon"
+                                                        !! [dim1_begin, dim1_end, dim2_begin, dim2_end,
+                                                        !!  dim3_begin, dim3_end, dim4_begin, dim4_end]
+     integer :: index_sub_region (NUM_SUB_REGION_ARRAY) !< Array that stores the index bounds for the sub region to
```

Comment:
It seems like a waste of memory to have these both on the stack if only one is going to be used.

Reply:
You need to have both. There is no real way around it. It can be allocatable but still ...

Reply:
Yes, you need both of them. But they should be `allocatable`, and only one should be allocated. Then you shouldn't need the `grid_type` variable because only one will be allocated (unless the string is used downstream for some reason).  
With this set up where they are both on the stack, you are just wasting memory.